Source/WebCore/bindings/js/WebCoreOpaqueRoot.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All Rights Reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#pragma once
+
+#include "Node.h"
+
+namespace WebCore {
+
+class WebCoreOpaqueRoot {
+public:
+    template<typename T>
+    explicit WebCoreOpaqueRoot(T* pointer)
+        : m_pointer(static_cast<void*>(pointer))
+        , m_isNode(pointer && std::is_base_of_v<Node, T>)
+    {
+        static_assert(!std::is_same_v<T, void>, "WebCoreOpaqueRoot cannot be constructed from a void*");
+    }
```

Comment:
I guess we could have done something like

typename = typename std::enable_if<!std::is_same_v<T, void>>::type

instead of static_assert?

Not sure that's necessarily better. Just trying to wrap my head around enable_if.

Reply:
I'll give your it a try. I can never remember how to use std::enable_if :)